src/curl.c

```diff
@@ -311,6 +311,99 @@ CURLcode swupd_download_file_complete(CURLcode curl_ret, struct curl_file *file)
 	return curl_ret;
 }
 
+
+// err is only needed because retry code is spread all over swupd. Remove this error code return when
+// #211 is fixed
+enum download_status process_curl_error_codes(int curl_ret, CURL *curl_handle, int *err)
+{
+	char *url;
+	if (curl_easy_getinfo(curl_handle, CURLINFO_EFFECTIVE_URL, &url) != CURLE_OK) {
+		url = "<not available>";
+	}
+
+	if (curl_ret == CURLE_OK || curl_ret == CURLE_HTTP_RETURNED_ERROR) {
+		long response = 0;
+		if (curl_ret == CURLE_OK || curl_ret == CURLE_HTTP_RETURNED_ERROR) {
+			if (curl_easy_getinfo(curl_handle, CURLINFO_RESPONSE_CODE, &response) != CURLE_OK) {
+				response = -1; // Force error
+			}
+		}
+		/* curl command succeeded, download might've failed, let our caller handle */
+		switch (response) {
+		case 206:
+			fprintf(stderr, "Curl: Partial file downloaded from '%s'\n", url);
+			*err = -1;
+			return DOWNLOAD_STATUS_PARTIAL_FILE;
+		case 200:
+		case 0:
+			*err = 0;
+			return DOWNLOAD_STATUS_COMPLETED;
+		case 403:
+			fprintf(stderr, "Curl: Download failed: Forbidden (403) - '%s'\n", url);
+			*err = -EPERM;
+			return DOWNLOAD_STATUS_FORBIDDEN;
+		case 404:
+			fprintf(stderr, "Curl: Download failed: File not found (403) - '%s'\n", url);
+			*err = -ENOENT;
+			return DOWNLOAD_STATUS_NOT_FOUND;
+		default:
+			*err = -ECOMM;
+			return DOWNLOAD_STATUS_ERROR;
+		}
+	} else { /* download failed but let our caller do it */
+		switch (curl_ret) {
+		case CURLE_COULDNT_RESOLVE_PROXY:
+			fprintf(stderr, "Curl: Could not resolve proxy\n");
+			*err = -1;
+			return DOWNLOAD_STATUS_ERROR;
+		case CURLE_COULDNT_RESOLVE_HOST:
+			fprintf(stderr, "Curl: Could not resolve host - '%s'\n", url);
+			*err = -1;
+			return DOWNLOAD_STATUS_ERROR;
+		case CURLE_COULDNT_CONNECT:
+			fprintf(stderr, "Curl: Could not connect to host or proxy\n");
```

Comment:
you can also print the url in this message similar to the way you do it in the CURLE_COULDNT_RESOLVE_HOST